### PR TITLE
[lazyinit] refactor lazy tensor and lazy init ctx

### LINKED ISSUE #3130: [lazyinit] Implement a single-process version of lazy init
## Overview

LazyTensor allows DL framework (PyTorch) to execute operations lazily, by storing all operations related to it and reruning them when it's required to be materialized.

LazyInit defers model initialiazation and it's based on LazyTensor.

This is essential when training large models.

Wanna track the development progress? Take a look at

proposal: #3124 
kanban: https://github.com/orgs/hpcaitech/projects/20

## Goal

The main entry point is `LazyInitContext`:

```python
ctx = LazyInitContext()
with ctx:
    model = GPT()

# if no need to distribute model
ctx.materialize(model)

# if need distribute model
   
sharding_strategy = symbolic_trace(model, ...)
ctx.distribute(model, sharding_strategy )
```

### PR DESCRIPTION
**Pull request by ver217**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

Closes #3130 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

Implement a single-process version of lazy initialization described in #3124 .

This code is tested locally, and as the model zoo is integrating, more tests will be added in the next PR.

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [ ] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/utils/model/experimental.py [hunk 1]
```diff
@@ -356,85 +343,50 @@ def wrapper(*args, **kwargs):
             setattr(torch, name, wrapper)
 
     def __exit__(self, exc_type, exc_val, exc_tb):
+        LazyInitContext._replaced = False
         for name, (wrapper, orig) in self.overrides.items():
             setattr(torch, name, orig)
 
     @staticmethod
-    def materialize(module: torch.nn.Module):
+    def materialize(module: torch.nn.Module, verbose: bool = False):
```
Comment (FrankLeeeee) [inline colossalai/utils/model/experimental.py:351]: `verbose` is not added to the docstring.

File: colossalai/utils/model/experimental.py [hunk 2]
```diff
@@ -50,140 +58,111 @@ class LazyTensor(torch.Tensor):
...
+                init_val = func(*tree_map(self._replace_with_materialized, args),
+                                **tree_map(self._replace_with_materialized, kwargs))
+            except TypeError as e:
+                print(f'init fn: {func.__name__}')
+                raise e
+
+            self._materialized_data = self._rerun_ops(init_val)
+        return self._materialized_data
+
+    def _rerun_ops(self, target=None) -> torch.Tensor:
```
Comment (FrankLeeeee) [inline colossalai/utils/model/experimental.py:148]: Add docstring to explain the definition of `rerun`.

File: colossalai/utils/model/experimental.py (hunk 1 above)
Comment (ver217) [reply colossalai/utils/model/experimental.py:351]: added

File: colossalai/utils/model/experimental.py (hunk 2 above)
Comment (ver217) [reply colossalai/utils/model/experimental.py:148]: done

**Review by FrankLeeeee — APPROVED** [review]

**Pull request merged**